File: quant/data/universe.py
"""Trading universe definitions for Indian equities."""

from __future__ import annotations

NIFTY50_SYMBOLS = {
    "ADANIENT.NS": "Adani Enterprises",
    "ADANIPORTS.NS": "Adani Ports",
    "APOLLOHOSP.NS": "Apollo Hospitals",
    "ASIANPAINT.NS": "Asian Paints",
    "AXISBANK.NS": "Axis Bank",
    "BAJAJ-AUTO.NS": "Bajaj Auto",
    "BAJFINANCE.NS": "Bajaj Finance",
    "BAJAJFINSV.NS": "Bajaj Finserv",
    "BPCL.NS": "Bharat Petroleum",
    "BHARTIARTL.NS": "Bharti Airtel",
    "BRITANNIA.NS": "Britannia",
    "CIPLA.NS": "Cipla",
    "COALINDIA.NS": "Coal India",
    "DIVISLAB.NS": "Divi's Labs",
    "DRREDDY.NS": "Dr. Reddy's",
    "EICHERMOT.NS": "Eicher Motors",
    "GRASIM.NS": "Grasim",
    "HCLTECH.NS": "HCL Tech",
    "HDFCBANK.NS": "HDFC Bank",
    "HDFCLIFE.NS": "HDFC Life",
    "HEROMOTOCO.NS": "Hero MotoCorp",
    "HINDALCO.NS": "Hindalco",
    "HINDUNILVR.NS": "Hindustan Unilever",
    "ICICIBANK.NS": "ICICI Bank",
    "INDUSINDBK.NS": "IndusInd Bank",
    "INFY.NS": "Infosys",
    "ITC.NS": "ITC",
    "JSWSTEEL.NS": "JSW Steel",
    "KOTAKBANK.NS": "Kotak Bank",
    "LT.NS": "Larsen & Toubro",
    "LTIM.NS": "LTIMindtree",
    "M&M.NS": "M&M",
    "MARUTI.NS": "Maruti Suzuki",
    "NESTLEIND.NS": "Nestle India",
    "NTPC.NS": "NTPC",
    "ONGC.NS": "ONGC",
    "POWERGRID.NS": "Power Grid",
    "RELIANCE.NS": "Reliance Industries",
    "SBILIFE.NS": "SBI Life",
    "SBIN.NS": "SBI",
    "SUNPHARMA.NS": "Sun Pharma",
    "TATACONSUM.NS": "Tata Consumer",
    # "TATAMOTORS.NS": "Tata Motors",  # yfinance symbol broken
    "TATASTEEL.NS": "Tata Steel",
    "TCS.NS": "TCS",
    "TECHM.NS": "Tech Mahindra",
    "TITAN.NS": "Titan",
    "ULTRACEMCO.NS": "UltraTech Cement",
    "UPL.NS": "UPL",
    "WIPRO.NS": "Wipro",
}

ETF = {
    "NIFTYBEES.NS": "Nifty 50 ETF",
    "GOLDBEES.NS": "Gold ETF",
    "SP500.NS": "S&P 500 ETF",
    "MIDCAPBEES.NS": "Midcap ETF",
    "SMLCAPBEES.NS": "Smallcap ETF",
    "LIQUIDBEES.NS": "Liquid ETF",
    "TATASILV.NS": "Tata Silver ETF",
    "SILVERBEES.NS": "Silver ETF",
    "KOTAKGOLD.NS": "Kotak Gold ETF",
    "SBIETFGOLD.NS": "SBI Gold ETF",
    "MON100.NS": "Nasdaq 100 ETF",
    "MAFANG.NS": "MAFANG ETF",
    "JUNIORBEES.NS": "Nifty Next 50 ETF",
    "BANKBEES.NS": "Bank Nifty ETF",
    "ITBEES.NS": "Nifty IT ETF",
    "PHARMABEES.NS": "Nifty Pharma ETF",
    "PSUBANKBEES.NS": "PSU Bank ETF",
}

INDEX_PROXY = {
    "^NSEI": "NIFTYBEES.NS",
    "^NSEBANK": "BANKBEES.NS",
    "^CNXIT": "ITBEES.NS",
}

STOCK = {
    **{s: NIFTY50_SYMBOLS[s] for s in NIFTY50_SYMBOLS},
    "IEX.NS": "IEX",
    "CDSL.NS": "CDSL",
    "ZOMATO.NS": "Zomato",
    "PAYTM.NS": "Paytm",
    "JIOFIN.NS": "Jio Financial",
    "DMART.NS": "DMart",
    "HAL.NS": "HAL",
    "VBL.NS": "Varun Beverages",
    "BSE.NS": "BSE Ltd",
    "ANGELONE.NS": "Angel One",
    "IRFC.NS": "IRFC",
    "RVNL.NS": "RVNL",
    "IREDA.NS": "IREDA",
}


def all_symbols() -> dict[str, str]:
    """Return all known symbols."""
    return {**STOCK, **ETF}


def is_india_symbol(sym: str) -> bool:
    """Check if symbol is Indian (NSE/BSE)."""
    return (
        sym.endswith(".NS")
        or sym.endswith(".BO")
        or sym.startswith("^NSE")
        or sym.startswith("^BSE")
    )


def normalize_symbol(sym: str) -> str:
    """Normalize symbol to yfinance format."""
    sym = sym.strip().upper()
    if sym in INDEX_PROXY:
        return INDEX_PROXY[sym]
    if sym.endswith((".NS", ".BO")):
        return sym
    if sym in ETF or sym in STOCK:
        return f"{sym}.NS"
    for s in STOCK:
        if s.replace(".NS", "") == sym:
            return s
    for s in ETF:
        if s.replace(".NS", "") == sym:
            return s
    return sym


def search_symbols(query: str) -> list[dict]:
    """Search symbols by name or ticker."""
    if not query:
        return []
    query = query.lower()
    items = all_symbols()
    return [
        {
            "symbol": s,
            "name": n,
            "exchange": "NSE",
            "type": "ETF" if s in ETF else "STOCK",
            "region": "India",
        }
        for s, n in items.items()
        if query in s.lower() or query in n.lower()
    ][:20]

File: web/lib/universe.ts
// Expanded stock universe - grouped by sector

export interface UniverseGroup {
  symbol: string
  label: string
  region: 'IN' | 'US' | 'UK' | 'CRYPTO'
}

export const UNIVERSE = {
  indices: [
    { symbol: '^NSEI',    label: 'NIFTY 50',   region: 'IN' as const },
    { symbol: '^BSESN',   label: 'SENSEX',      region: 'IN' as const },
    { symbol: '^NSEBANK', label: 'BANKNIFTY',   region: 'IN' as const },
    { symbol: '^NSMIDCP', label: 'MIDCAP',      region: 'IN' as const },
    { symbol: '^GSPC',    label: 'S&P 500',     region: 'US' as const },
    { symbol: '^DJI',     label: 'DOW',         region: 'US' as const },
    { symbol: '^IXIC',    label: 'NASDAQ',      region: 'US' as const },
    { symbol: '^FTSE',    label: 'FTSE 100',    region: 'UK' as const },
  ],
  nifty50_it: [
    'TCS.NS', 'INFY.NS', 'HCLTECH.NS', 'WIPRO.NS', 'TECHM.NS', 'LTIM.NS'
  ],
  nifty50_banking: [
    'HDFCBANK.NS', 'ICICIBANK.NS', 'KOTAKBANK.NS', 'SBIN.NS', 
    'AXISBANK.NS', 'INDUSINDBK.NS', 'BANDHANBNK.NS'
  ],
  nifty50_energy: [
    'RELIANCE.NS', 'ONGC.NS', 'BPCL.NS', 'NTPC.NS', 'POWERGRID.NS', 'COALINDIA.NS'
  ],
  nifty50_pharma: [
    'SUNPHARMA.NS', 'DRREDDY.NS', 'CIPLA.NS', 'DIVISLAB.NS', 'APOLLOHOSP.NS'
  ],
  nifty50_auto: [
    'MARUTI.NS', 'TATAMOTORS.NS', 'BAJAJ-AUTO.NS', 'HEROMOTOCO.NS', 'EICHERMOT.NS', 'M&M.NS'
  ],
  nifty50_fmcg: [
    'HINDUNILVR.NS', 'ITC.NS', 'NESTLEIND.NS', 'BRITANNIA.NS', 'TATACONSUM.NS'
  ],
  nifty50_metals: [
    'TATASTEEL.NS', 'HINDALCO.NS', 'JSWSTEEL.NS', 'COALINDIA.NS'
  ],
  global: [
    'AAPL', 'MSFT', 'GOOGL', 'AMZN', 'NVDA', 'META', 'TSLA'
  ],
  crypto: [
    'BTC-USD', 'ETH-USD', 'SOL-USD', 'BNB-USD'
  ],
}

// Symbol to company name mapping for news filtering
export const COMPANY_NAMES: Record<string, string> = {
  'RELIANCE.NS': 'Reliance Industries',
  'TCS.NS': 'Tata Consultancy Services',
  'HDFCBANK.NS': 'HDFC Bank',
  'INFY.NS': 'Infosys',
  'ICICIBANK.NS': 'ICICI Bank',
  'SBIN.NS': 'State Bank of India',
  'HINDUNILVR.NS': 'Hindustan Unilever',
  'ITC.NS': 'ITC Limited',
  'TATAMOTORS.NS': 'Tata Motors',
  'MARUTI.NS': 'Maruti Suzuki',
  'SUNPHARMA.NS': 'Sun Pharmaceutical',
  'TATASTEEL.NS': 'Tata Steel',
  'WIPRO.NS': 'Wipro Limited',
  'AXISBANK.NS': 'Axis Bank',
  'KOTAKBANK.NS': 'Kotak Mahindra Bank',
  'NESTLEIND.NS': 'Nestle India',
  'ONGC.NS': 'Oil and Natural Gas Corporation',
  'POWERGRID.NS': 'Power Grid Corporation',
  'NTPC.NS': 'NTPC Limited',
  'COALINDIA.NS': 'Coal India',
  'DRREDDY.NS': 'Dr Reddy\'s Laboratories',
  'CIPLA.NS': 'Cipla Limited',
  'DIVISLAB.NS': 'Divi\'s Laboratories',
  'APOLLOHOSP.NS': 'Apollo Hospitals',
  'BAJAJ-AUTO.NS': 'Bajaj Auto',
  'HEROMOTOCO.NS': 'Hero MotoCorp',
  'EICHERMOT.NS': 'Eicher Motors',
  'M&M.NS': 'Mahindra & Mahindra',
  'BRITANNIA.NS': 'Britannia Industries',
  'TATACONSUM.NS': 'Tata Consumer Products',
  'HINDALCO.NS': 'Hindalco Industries',
  'JSWSTEEL.NS': 'JSW Steel',
  'TECHM.NS': 'Tech Mahindra',
  'LTIM.NS': 'L&T Infotech',
  'HCLTECH.NS': 'HCL Technologies',
  'INDUSINDBK.NS': 'IndusInd Bank',
  'BANDHANBNK.NS': 'Bandhan Bank',
  'NSEI': 'Nifty 50',
  'BSESN': 'BSE Sensex',
  'NSEBANK': 'Nifty Bank',
  '^NSEI': 'Nifty 50',
  '^BSESN': 'BSE Sensex',
  '^NSEBANK': 'Nifty Bank',
  '^GSPC': 'S&P 500',
  '^DJI': 'Dow Jones',
  '^IXIC': 'NASDAQ',
  'AAPL': 'Apple Inc',
  'MSFT': 'Microsoft',
  'GOOGL': 'Alphabet',
  'AMZN': 'Amazon',
  'NVDA': 'NVIDIA',
  'META': 'Meta Platforms',
  'TSLA': 'Tesla',
  'BTC-USD': 'Bitcoin',
  'ETH-USD': 'Ethereum',
  'SOL-USD': 'Solana',
  'BNB-USD': 'Binance Coin',
}

// Get company name for a symbol
export function getCompanyName(symbol: string): string {
  return COMPANY_NAMES[symbol] || symbol.replace('.NS', '').replace('-USD', '')
}

// Get all symbols flattened
export function getAllSymbols(): string[] {
  return [
    ...UNIVERSE.indices.map(i => i.symbol),
    ...UNIVERSE.nifty50_it,
    ...UNIVERSE.nifty50_banking,
    ...UNIVERSE.nifty50_energy,
    ...UNIVERSE.nifty50_pharma,
    ...UNIVERSE.nifty50_auto,
    ...UNIVERSE.nifty50_fmcg,
    ...UNIVERSE.nifty50_metals,
    ...UNIVERSE.global,
    ...UNIVERSE.crypto,
  ]
}

// Default watchlist symbols
export const DEFAULT_WATCHLIST = [
  'RELIANCE.NS',
  'TCS.NS', 
  'HDFCBANK.NS',
  'INFY.NS',
  'ICICIBANK.NS',
  'SBIN.NS',
  '^NSEI',
]
